quantestpy/simulator/pauli_circuit.py

```diff
@@ -0,0 +1,143 @@
+import numpy as np
+
+from quantestpy.exceptions import QuantestPyError, QuantestPyTestCircuitError
+from quantestpy.test_circuit import TestCircuit
+
+_IMPLEMENTED_GATES = ["x", "y", "z", "swap"]
+
+
+class PauliCircuit(TestCircuit):
+
+    def __init__(self, num_qubit: int):
+        if not isinstance(num_qubit, int) or num_qubit < 1:
+            raise QuantestPyTestCircuitError(
+                "num_qubit must be an integer greater than 0."
+            )
+
+        super().__init__(num_qubit=num_qubit)
+        self._qubit_value = np.array([0 for _ in range(num_qubit)])
+        self._qubit_phase = np.array([0. for _ in range(num_qubit)])
+
+    def add_gate(self, gate: dict) -> None:
+        if not isinstance(gate, dict):
+            raise QuantestPyTestCircuitError(
+                "gate must be a dictionary which contains 'name', "
+                "'target_qubit', 'control_qubit' and 'control_value' as keys."
+            )
+
+        if "name" not in gate.keys():
+            raise QuantestPyTestCircuitError(
+                "gate must contain 'name' as a key."
+            )
+
+        if gate["name"] not in _IMPLEMENTED_GATES:
+            raise QuantestPyTestCircuitError(
+                f'{gate["name"]} gate is not implemented.\n'
+                f'Implemented gates: {_IMPLEMENTED_GATES}'
+            )
+
+        gate["parameter"] = []  # to avoid raise in TestCircuit.add_gate()
+        super().add_gate(gate=gate)
+
+    @staticmethod
+    def _assert_is_pauli_circuit(circuit, circuit_name: str):
+        if not isinstance(circuit, PauliCircuit):
+            raise QuantestPyTestCircuitError(
+                f"{circuit_name} must be an instance of PauliCircuit class."
+            )
+
+    def _assert_is_correct_reg(self, reg, reg_name: str):
+        if not isinstance(reg, list):
+            raise QuantestPyTestCircuitError(f"{reg_name} must be a list.")
+
+        for idx in reg:
+            if not isinstance(idx, int):
+                raise QuantestPyTestCircuitError(
+                    f"Indices in {reg_name} must be integer type."
+                )
+            if idx >= self._num_qubit or idx < 0:
+                raise QuantestPyTestCircuitError(
+                    f"Qubit index {idx} in {reg_name} is out of range."
+                )
+
+    @staticmethod
```

Comment:
No it does not have to be a staticmethod but it can be, so I make it a staticmethod. This is often suggested for the purpose of readability as long as I know.